src/client/pmix_client.c

```diff
@@ -570,56 +561,81 @@ pmix_status_t PMIx_Put(pmix_scope_t scope, const char key[], pmix_value_t *val)
 
     /* pack the cache that matches the scope - global scope needs
      * to go into both local and remote caches */
-    if (PMIX_LOCAL == scope || PMIX_GLOBAL == scope) {
+    if (PMIX_LOCAL == cb->scope || PMIX_GLOBAL == cb->scope) {
         if (NULL == pmix_globals.cache_local) {
             pmix_globals.cache_local = PMIX_NEW(pmix_buffer_t);
         }
         pmix_output_verbose(2, pmix_globals.debug_output,
                             "pmix: put %s data for key %s in local cache",
-                            key, (PMIX_GLOBAL == scope) ? "global" : "local");
+                            cb->key, (PMIX_GLOBAL == cb->scope) ? "global" : "local");
         if (PMIX_SUCCESS != (rc = pmix_bfrop.pack(pmix_globals.cache_local, kv, 1, PMIX_KVAL))) {
             PMIX_ERROR_LOG(rc);
         }
     }
 
-    if (PMIX_REMOTE == scope || PMIX_GLOBAL == scope) {
+    if (PMIX_REMOTE == cb->scope || PMIX_GLOBAL == cb->scope) {
         if (NULL == pmix_globals.cache_remote) {
             pmix_globals.cache_remote = PMIX_NEW(pmix_buffer_t);
         }
         pmix_output_verbose(2, pmix_globals.debug_output,
                             "pmix: put %s data for key %s in remote cache",
-                            key, (PMIX_GLOBAL == scope) ? "global" : "remote");
+                            cb->key, (PMIX_GLOBAL == cb->scope) ? "global" : "remote");
         if (PMIX_SUCCESS != (rc = pmix_bfrop.pack(pmix_globals.cache_remote, kv, 1, PMIX_KVAL))) {
             PMIX_ERROR_LOG(rc);
         }
     }
 
+  done:
     PMIX_RELEASE(kv);  // maintain accounting
+    cb->pstatus = rc;
+    cb->active = false;
+}
+
+pmix_status_t PMIx_Put(pmix_scope_t scope, const char key[], pmix_value_t *val)
+{
+    pmix_cb_t *cb;
+    pmix_status_t rc;
+
+    pmix_output_verbose(2, pmix_globals.debug_output,
+                        "pmix: executing put for key %s type %d",
+                        key, val->type);
+
+    if (pmix_globals.init_cntr <= 0) {
+        return PMIX_ERR_INIT;
+    }
+
+    /* create a callback object */
+    cb = PMIX_NEW(pmix_cb_t);
+    cb->active = true;
+    cb->scope = scope;
+    cb->key = (char*)key;
```

Comment:
@artpol84 I have no issue with doing extra strdup's - I didn't include it because you had later flagged lines for having strdup's that weren't needed because we could NULL the variable prior to release. At some point, we need to decide which way we are going to go and be consistent.

Adding strdup's is indeed safer and protects against any future case where the data isn't statically passed to us. There is a performance penalty to all these tiny malloc's since we don't have a memory manager in the code like OMPI does, but it probably isn't significant - modern Linux implementations have gotten fairly efficient at malloc.

The only issue would be that we definitely don't want to copy all those info arrays and proc arrays! So we will be treating the string values different from any other passed value. This is a little troubling.

Other option is to remove the free's from the pmix_cb_t destructor, and require that people take care of it directly in the code. It might result in leaks, but those can be stomped out. Frankly, I'm thinking this is the correct path.

Thoughts?